lib/bundler/definition.rb

```diff
@@ -800,10 +800,12 @@ def converge_locked_specs
           # commonly happens if the version changed in the gemspec
           next unless other
 
-          deps2 = other.dependencies.select {|d| d.type != :development }
-          runtime_dependencies = s.dependencies.select {|d| d.type != :development }
-          # If the dependencies of the path source have changed, unlock it
-          next unless runtime_dependencies.sort == deps2.sort
+          new_deps = other.dependencies.select {|d| d.type != :development }
+          old_deps = s.dependencies.select {|d| d.type != :development }
+          # If the dependencies of the path source have changed and locked spec can't satisfy new dependencies, unlock it
+          next unless old_deps.sort == new_deps.sort || new_deps.all? {|d| satisfies_locked_spec?(d) }
+
+          s.dependencies.replace(other.dependencies)
```

Comment:
I don't think this line is necessary, isn't it enough to add the condition about the new dependencies satisfying the locked spec?

Also, I would split the variable rename, which is great because it makes the code more readable, to a separate commit.

Reply:
> I don't think this line is necessary, isn't it enough to add the condition about the new dependencies satisfying the locked spec?

Without this line, Gemfile.lock won't be updated.

<details>
<summary>result of ` bin/rspec spec/install/gemfile/path_spec.rb:488` with patch</summary>

### Patch

```diff
diff --git a/lib/bundler/definition.rb b/lib/bundler/definition.rb
index 81b362fed..4bf4afb47 100644
--- a/lib/bundler/definition.rb
+++ b/lib/bundler/definition.rb
@@ -805,7 +805,7 @@ module Bundler
           # If the dependencies of the path source have changed and locked spec can't satisfy new dependencies, unlock it
           next unless deps2.sort == runtime_dependencies.sort || deps2.all? {|d| satisfies_locked_spec?(d) }
 
-          s.dependencies.replace(other.dependencies)
+          # s.dependencies.replace(other.dependencies)
         end
 
         converged << s
```

### Result

```
app@967f4685ddfc:/app$ bin/rspec spec/install/gemfile/path_spec.rb:488
Run options:
  include {:locations=>{"./spec/install/gemfile/path_spec.rb"=>[488]}}
  exclude {:ruby_repo=>false, :bundler=>#<RequirementChecker:./spec/support/requirement_checker.rb:7>, :rubygems_master=>true, :rubygems=>#<RequirementChecker:./spec/support/requirement_checker.rb:7>, :ruby=>#<RequirementChecker:./spec/support/requirement_checker.rb:7>, :realworld=>true, :sudo=>true}

bundle install with explicit source paths
  when dependencies in the path are updated
$ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
  /app/exe/bundle install --retry 0
Fetching source index from file:///app/tmp/gems/remote1/
Resolving dependencies...
Using bundler 2.1.0.pre.1
Using foo 1.0 from source at `/app/tmp/libs/foo`
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
# $? => 0

$ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
  /app/exe/bundle install
Fetching source index from file:///app/tmp/gems/remote1/
Resolving dependencies...
Using bundler 2.1.0.pre.1
Fetching rack 0.9.1
Installing rack 0.9.1
Using foo 1.0 from source at `/app/tmp/libs/foo`
Bundle complete! 1 Gemfile dependency, 3 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Post-install message from rack:
Rack's post install message
# $? => 0

$ /usr/local/bin/ruby -I/app/lib -e  \
  <<EOS
  require 'bundler' ; Bundler.setup()
  require 'rack.rb'; puts RACK
EOS
0.9.1
# $? => 0

$ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
  /app/exe/bundle install
Fetching source index from file:///app/tmp/gems/remote1/
Resolving dependencies...
Using bundler 2.1.0.pre.1
Using rack 0.9.1
Using foo 1.0 from source at `/app/tmp/libs/foo`
Bundle complete! 1 Gemfile dependency, 3 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
# $? => 0
    keeps using the same version if it's compatible (FAILED - 1)


Retried examples: 0

Failures:

  1) bundle install with explicit source paths when dependencies in the path are updated keeps using the same version if it's compatible
     Failure/Error: expect(bundled_app("Gemfile.lock")).to have_lockfile(expected)
     
       expected "PATH\n  remote: /app/tmp/libs/foo\n  specs:\n    foo (1.0)\n      rack (= 0.9.1)\n\nGEM\n  remote: f...s:\n    rack (0.9.1)\n\nPLATFORMS\n  ruby\n\nDEPENDENCIES\n  foo!\n\nBUNDLED WITH\n   2.1.0.pre.1\n" to read as "PATH\n  remote: /app/tmp/libs/foo\n  specs:\n    foo (1.0)\n      rack\n\nGEM\n  remote: file:///app...s:\n    rack (0.9.1)\n\nPLATFORMS\n  ruby\n\nDEPENDENCIES\n  foo!\n\nBUNDLED WITH\n   2.1.0.pre.1\n"
       Diff:
       @@ -2,7 +2,7 @@
          remote: /app/tmp/libs/foo
          specs:
            foo (1.0)
       -      rack
       +      rack (= 0.9.1)
        
        GEM
          remote: file:///app/tmp/gems/remote1/
       
     
       Commands:
       $ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
         /app/exe/bundle install --retry 0
       Fetching source index from file:///app/tmp/gems/remote1/
       Resolving dependencies...
       Using bundler 2.1.0.pre.1
       Using foo 1.0 from source at `/app/tmp/libs/foo`
       Bundle complete! 1 Gemfile dependency, 2 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       # $? => 0
     
       $ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
         /app/exe/bundle install
       Fetching source index from file:///app/tmp/gems/remote1/
       Resolving dependencies...
       Using bundler 2.1.0.pre.1
       Fetching rack 0.9.1
       Installing rack 0.9.1
       Using foo 1.0 from source at `/app/tmp/libs/foo`
       Bundle complete! 1 Gemfile dependency, 3 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       Post-install message from rack:
       Rack's post install message
       # $? => 0
     
       $ /usr/local/bin/ruby -I/app/lib -e  \
         <<EOS
         require 'bundler' ; Bundler.setup()
         require 'rack.rb'; puts RACK
       EOS
       0.9.1
       # $? => 0
     
       $ /usr/local/bin/ruby -I/app/lib:/app/spec -rsupport/hax -r/app/spec/support/artifice/fail \
         /app/exe/bundle install
       Fetching source index from file:///app/tmp/gems/remote1/
       Resolving dependencies...
       Using bundler 2.1.0.pre.1
       Using rack 0.9.1
       Using foo 1.0 from source at `/app/tmp/libs/foo`
       Bundle complete! 1 Gemfile dependency, 3 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       # $? => 0
     # ./spec/support/matchers.rb:215:in `lockfile_should_be'
     # ./spec/install/gemfile/path_spec.rb:525:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:98:in `block (2 levels) in <top (required)>'

Finished in 4.13 seconds (files took 0.53702 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/install/gemfile/path_spec.rb:488 # bundle install with explicit source paths when dependencies in the path are updated keeps using the same version if it's compatible
```

</details>

> Also, I would split the variable rename, which is great because it makes the code more readable, to a separate commit.

Ok, I'll send another PR for that.

Reply:
Interesting, I did try reverting the line and the spec passed for me, maybe I did it wrong... :thinking:. I'll double check again.

> Ok, I'll send another PR for that.

A separate commit works too, whatever you prefer!

Reply:
> I noticed added spec isn't enough so I added some assertions to make sure Gemfile.lock is updated.

Ah ok, I missed this part in your message. That's great, thanks for updating!

Reply:
> Interesting, I did try reverting the line and the spec passed for me, maybe I did it wrong... 🤔.

Sorry, my added spec is not enough at the first time 🙇 
I guess you tried on old revision